liblwgeom/lwgeom_geos.c

```diff
@@ -273,18 +274,24 @@ ptarray_to_GEOSLinearRing(const POINTARRAY *pa, int autofix)
 		*/
 		if ( ! ptarray_is_closed_2d(pa) || pa->npoints < 4 )
 		{
-			pa = ptarray_addPoint(pa, getPoint_internal(pa, 0), FLAGS_NDIMS(pa->flags), pa->npoints);
-			npa = pa;
+			npa = ptarray_addPoint(pa, getPoint_internal(pa, 0), FLAGS_NDIMS(pa->flags), pa->npoints);
 		}
 		/* Check ring for having at least 4 vertices */
-		while ( pa->npoints < 4 )
+		while ( npa->npoints < 4 )
 		{
-			ptarray_append_point(pa, getPoint_internal(pa, 0), LW_TRUE);
+			ptarray_append_point(npa, (POINT4D*)getPoint_internal(npa, 0), LW_TRUE);
```

Comment:
Would you use getPoint4d_p instead? Getting addicted to pointing right into the array is a good way to start reading off the end of it.